pkg/controller/workload/workload_controller.go

```diff
@@ -46,14 +46,34 @@ func NewController(workloadMap bpf2go.KmeshCgroupSockWorkloadMaps) *Controller {
 }
 
 func (ws *Controller) WorkloadStreamCreateAndSend(client discoveryv3.AggregatedDiscoveryServiceClient, ctx context.Context) error {
-	var err error
+	var (
+		err                  error
+		addressResourceNames []string
+	)
 
 	ws.Stream, err = client.DeltaAggregatedResources(ctx)
 	if err != nil {
 		return fmt.Errorf("DeltaAggregatedResources failed, %s", err)
 	}
 
-	if err := ws.Stream.Send(newWorkloadRequest(AddressType, nil)); err != nil {
+	if ws.Processor != nil {
+		cachedServices := ws.Processor.ServiceCache.List()
+		cachedWorkloads := ws.Processor.WorkloadCache.List()
+		addressResourceNames = make([]string, 0, len(cachedServices)+len(cachedWorkloads))
+
+		// add cached resource names
+		for _, service := range cachedServices {
+			addressResourceNames = append(addressResourceNames, service.ResourceName())
+		}
+
+		for _, workload := range cachedWorkloads {
+			addressResourceNames = append(addressResourceNames, workload.ResourceName())
+		}
+	}
+
+	log.Debugf("Initial address resource names: %v", addressResourceNames)
+
+	if err := ws.Stream.Send(newWorkloadRequest(AddressType, addressResourceNames)); err != nil {
```

Comment:
This is not right, you should set the `InitialResourceVersions`

Reply:
For the versions, we can set empty "", because the server side does not check them now

Reply:
right, `ResourceNamesSubscribe` unset means client is subscribing all the resources.

Only for on demand xds this can be set to specific values